distributed/s3.py

```diff
@@ -26,148 +32,635 @@
+                try:
+                    files = [self.info(path)]
+                except OSError:
+                    files = []
+        if detail:
+            return files
+        else:
+            return [f['Key'] for f in files]
+    
+    def info(self, path):
+        path = path.lstrip('s3://').rstrip('/')
+        files = self._ls(path)
+        files = [f for f in files if f['Key'].rstrip('/') == path]
+        if len(files) == 1:
+            return files[0]
+        else:
+            raise IOError("File not found: %s" %path)
+
+    def walk(self, path):
+        return [f['Key'] for f in self._ls(path) if f['Key'].rstrip('/'
+                ).startswith(path.rstrip('/') + '/')]
+
+    def glob(self, path):
+        """
+        Find files by glob-matching.
+
+        Note that the bucket part of the path must not contain a "*"
+        """
+        path0 = path
+        path = path.lstrip('s3://').lstrip('/')
+        bucket, key = split_path(path)
+        if "*" in bucket:
+            raise ValueError('Bucket cannot contain a "*"')
+        if '*' not in path:
+            path = path.rstrip('/') + '/*'
+        if '/' in path[:path.index('*')]:
+            ind = path[:path.index('*')].rindex('/')
+            root = path[:ind+1]
+        else:
+            root = '/'
+        allfiles = self.walk(root)
+        pattern = re.compile("^" + path.replace('//', '/')
+                                        .rstrip('/')
+                                        .replace('*', '[^/]*')
+                                        .replace('?', '.') + "$")
+        out = [f for f in allfiles if re.match(pattern,
+               f.replace('//', '/').rstrip('/'))]
+        if not out:
+            out = self.ls(path0)
+        return out
+
+    def du(self, path, total=False, deep=False):
+        if deep:
+            files = self.walk(path)
+            files = [self.info(f) for f in files]
+        else:
+            files = self.ls(path, detail=True)
+        if total:
+            return sum(f.get('Size', 0) for f in files)
+        else:
+            return {p['Key']: p['Size'] for p in files}
+
+    def df(self):
+        total = 0
+        for files in self.dirs.values():
+            total += sum(f['Size'] for f in files)
+        return {'capacity': None, 'used': total, 'percent-free': None}
+
+    def __repr__(self):
+        return 'S3 File System'
+
+    def mkdir(self, path):
+        raise NotImplementedError('S3 implemented as read-only')
+
+    def touch(self, path):
+        raise NotImplementedError('S3 implemented as read-only')
+
+    def mv(self, path1, path2):
+        raise NotImplementedError('S3 implemented as read-only')
+
+    def rm(self, path, recursive=True):
+        raise NotImplementedError('S3 implemented as read-only')
+
+    def exists(self, path):
+        return bool(self.ls(path))
+
+    def copy(self, path1, path2):
+        raise NotImplementedError('S3 implemented as read-only')
```

Comment:
I would prefer to remove these from the interface for now.  Otherwise we're just setting users up for disappointment.